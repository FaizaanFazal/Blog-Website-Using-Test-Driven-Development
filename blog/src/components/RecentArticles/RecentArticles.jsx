import React from 'react'
import Card from './Card'
import {useSelector} from 'react-redux'

export default function RecentArticles() {

   const recentArticles =useSelector(state=>state.blog.recentArticles);

  return (
    <section className="recent-sc">
    <div className="container">
      <div className="recent-content">
        <div className="section-head flex items-end justify-between">
          <div className="sc-head-l">
            <h4 className="title title-lg">Recent Articles</h4>
            <p className="text text-base">
              Here&apos;s what we&apos;ve been up to recently.
            </p>
          </div>
          <div className="sc-head-r">
            <button type="button" className="btn btn-o-primary">
              View all
              <span className="btn-icon">
                <i className="bi bi-arrow-right"></i>
              </span>
            </button>
          </div>
        </div>

        <div className="card-list grid-cols grid-cols-3">
          {
            recentArticles?.map(blogItem => {
              return (<Card blogItemData = {blogItem} isCardSm = { true } key = {blogItem.id} />)
            })
          }
        </div>
      </div>
    </div>
  </section>
  )
}
