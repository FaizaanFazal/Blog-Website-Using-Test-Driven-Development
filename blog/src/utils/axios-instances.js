// import axios from 'axios';

// export const axiosInstance = axios.create({
//   baseURL: 'http://localhost:8000/',
// });
import axios from 'axios';

const placeholderApi = axios.create({
  baseURL: 'http://localhost:8000',
});

export default placeholderApi;
