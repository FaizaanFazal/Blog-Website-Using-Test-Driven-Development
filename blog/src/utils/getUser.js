export const getUser = () => {
    const userString = localStorage.getItem('user');
    if (userString) {
      const user = JSON.parse(userString);
      return user;
    } else {
      return null;
    }
  };