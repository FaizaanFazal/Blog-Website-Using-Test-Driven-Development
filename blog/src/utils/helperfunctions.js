import { fireEvent } from '@testing-library/react';
import images from './images';

export const isProperImageURL = (url) => {
  /* eslint-disable-next-line no-useless-escape */
  const regex = /^(https?|ftp):\/\/(([a-z\d]([a-z\d-]*[a-z\d])?\.)+[a-z]{2,}|localhost)(\/[-a-z\d%_.~+]*)*(\?[;&a-z\d%_.~+=-]*)?(\#[-a-z\d_]*)?$/i;
  return regex.test(url);
};
export const isProperEmail = (email) => {
  const emailRegex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return emailRegex.test(email);
};
export const isProperPass = (pass) => {
  const passregex = /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[@$!%*#?&])[A-Za-z\d@$!%*#?&]{8,}$/;
  return passregex.test(pass);
};
export function backspace(element) {
  let actuallyTyped = element.value;

  const backspaceKey = {
    key: 'Backspace',
    code: 8,
    inputType: 'deleteContentBackward',
  };

  const sharedEventConfig = {
    key: backspaceKey.key,
    charCode: backspaceKey.code,
    keyCode: backspaceKey.code,
    which: backspaceKey.code,
    modifier: backspaceKey.modifier,
  };
  const downEvent = fireEvent.keyDown(element, sharedEventConfig);

  if (downEvent) {
    actuallyTyped = actuallyTyped.slice(0, -1);

    fireEvent.input(element, {
      target: { value: actuallyTyped },
      inputType: backspaceKey.inputType,
      bubbles: true,
      cancelable: true,
    });
  }

  fireEvent.keyUp(element, sharedEventConfig);
}

export const small = [
  {
    id: 1,
    title: '10 Tips for Budget-Friendly Travel',
    author: 'Elena Rodriguez',
    authorImage: { src: images.author1, alt: 'Elena Rodriguez' },
    date: '15 August',
    image: { src: images.blogImg1, alt: '10 Tips for Budget-Friendly Travel Blog Header Image' },
    verified: true,
    slug: 'post1',
    content: "test Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.",
  },
];
export const greater = [
  {
    id: 1,
    title: '10 Tips for Budget-Friendly Travel',
    author: 'Elena Rodriguez',
    authorImage: { src: images.author1, alt: ' Elena' },
    date: '15 August',
    image: { src: images.blogImg1, alt: '10 Tips for Budget-Friendly Travel Blog Header Image' },
    slug: 'post1',
    verified: true,
    content: "Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank. \n \n Traveling on a budget doesnt have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank. \n \n Traveling on a budget doesnt have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank. \n",
  }];
