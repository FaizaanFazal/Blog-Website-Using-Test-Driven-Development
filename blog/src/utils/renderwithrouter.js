import {render} from '@testing-library/react'
import * as userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom'
import {MemoryRouter} from 'react-router-dom'

export const renderWithRouter = (ui, {route = '/'} = {}) => {
  window.history.pushState({}, 'Test page', route)

  return {
    user: userEvent.setup,
    ...render(ui, {wrapper: MemoryRouter}),
  }
}