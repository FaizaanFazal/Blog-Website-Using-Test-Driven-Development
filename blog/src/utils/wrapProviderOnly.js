import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import PropTypes from 'prop-types';
import reduxStore from '../Redux/store';
import '@testing-library/jest-dom';

export * from '@testing-library/react';

const AllTheProviders = ({ children }) => (
  <Provider store={reduxStore}>
    {children}
  </Provider>
);

export const renderWithProvidersOnly = (ui, options) => render(<AllTheProviders>{ui}</AllTheProviders>, options);

// export function renderWithProviders(
//   ui,
//   {
//     // preloadedState = {},
//     // Automatically create a store instance if no store was passed in
//     store = reduxStore,
//     ...renderOptions
//   } = {},
// ) {
//   function Wrapper({ children }) {
//     return <Provider store={store}>{children}</Provider>;
//   }

//   // Return an object with the store and all of RTL's query functions
//   return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
// }

renderWithProvidersOnly.propTypes = {
  children: PropTypes.node.isRequired,
};
