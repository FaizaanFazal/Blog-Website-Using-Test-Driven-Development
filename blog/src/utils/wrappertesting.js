import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import PropTypes from 'prop-types';
import reduxStore from '../Redux/store';
import '@testing-library/jest-dom';

export * from '@testing-library/react';

export function renderWithProviders(
  ui,
  {
    // preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = reduxStore,
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

renderWithProviders.propTypes = {
  children: PropTypes.node.isRequired,
};
