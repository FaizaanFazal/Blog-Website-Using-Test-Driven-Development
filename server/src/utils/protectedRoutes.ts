import { authenticateToken } from "@/models/authModel";
import { Request, Response, NextFunction, RequestHandler } from "express";


export const protectedRoutes = [
    {
        method: "",
        path: ""
    },
    {
        method: "",
        path: ""
    }];


//check protected routes to check authentication
export const isProtectedRouteMiddleware: RequestHandler = (req: Request, res: Response, next: NextFunction) => {
    console.log(req.method)
    const protectedPathData = protectedRoutes.find((route) => route.path === req.path);
    if (protectedPathData) {
        const authenticatedUser = authenticateToken(req, res, next);
        if (authenticatedUser !== 403 && authenticatedUser !== 401) {
            req.user = authenticatedUser;
            console.log("User", authenticatedUser)
        } else {
            res.status(authenticatedUser);
            res.end();
        }
    } else {
        next();
    }
};