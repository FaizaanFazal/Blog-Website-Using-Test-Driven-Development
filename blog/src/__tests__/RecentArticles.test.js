import React from 'react';
import { screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import RecentArticles from '../components/RecentArticles/RecentArticles';
import { renderWithProviders } from '../utils/wrappertesting';


describe('Unit tests for RecentArticle Components', () => {
  it('Snapshot test for Recent Article component', () => {
    const { asFragment } = renderWithProviders(<MemoryRouter><RecentArticles /></MemoryRouter>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Number of Articles should be 3', () => {
    renderWithProviders(<MemoryRouter><RecentArticles /></MemoryRouter>);
    const tagElements = screen.getAllByTestId('artice');
    expect(tagElements.length).toBe(3);
  });

  it('View All Button', () => {
    renderWithProviders(<MemoryRouter><RecentArticles /></MemoryRouter>);
    const searchbtn = screen.getByText('View all');
    expect(searchbtn).toBeInTheDocument();
  });
});
