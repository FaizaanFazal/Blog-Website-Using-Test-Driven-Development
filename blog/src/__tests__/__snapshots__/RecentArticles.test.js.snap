// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for RecentArticle Components Snapshot test for Recent Article component 1`] = `
<DocumentFragment>
  <section
    class="recent-sc"
  >
    <div
      class="container"
    >
      <div
        class="recent-content"
      >
        <div
          class="section-head flex items-end justify-between"
        >
          <div
            class="sc-head-l"
          >
            <h4
              class="title title-lg"
            >
              Recent Articles
            </h4>
            <p
              class="text text-base"
            >
              Here's what we've been up to recently.
            </p>
          </div>
          <div
            class="sc-head-r"
          >
            <button
              class="btn btn-o-primary"
              type="button"
            >
              View all
              <span
                class="btn-icon"
              >
                <i
                  class="bi bi-arrow-right"
                />
              </span>
            </button>
          </div>
        </div>
        <div
          class="card-list grid-cols grid-cols-3"
        >
          <a
            data-testid="slugLink"
            href="/blogs/post2"
          >
            <div
              class="card card-sm"
              data-testid="artice"
            >
              <img
                alt="Article"
                class="card-overlay-img object-fit-cover"
                data-testid="Image"
                src="blogImg2.jpg"
              />
              <div
                class="card-content"
              >
                <div
                  class="card-body"
                >
                  <h3
                    class="title title-base"
                  >
                    The Importance of Home Safety Measures
                  </h3>
                </div>
                <div
                  class="card-footer card-footer-light flex justify-between items-center"
                >
                  <div
                    class="writer-info grid"
                  >
                    <div
                      class="info-avatar"
                    >
                      <img
                        alt=""
                        class="object-fit-cover"
                        src="author2.jpg"
                      />
                    </div>
                    <div
                      class="info-intro"
                    >
                      <p
                        class="intro-name text-base"
                      >
                        Jonathan Baker
                      </p>
                      <div
                        class="intro-verify flex items-center"
                      >
                        <img
                          alt="icon"
                          class="verify-icon"
                          src="verify_icon_filled.svg"
                        />
                        <span
                          class="verify-status text-sm text"
                        >
                          Verified writer
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    class="date-info text text-base"
                  >
                    18 August
                  </div>
                </div>
              </div>
            </div>
          </a>
          <a
            data-testid="slugLink"
            href="/blogs/post3"
          >
            <div
              class="card card-sm"
              data-testid="artice"
            >
              <img
                alt="Article"
                class="card-overlay-img object-fit-cover"
                data-testid="Image"
                src="blogImg3.jpg"
              />
              <div
                class="card-content"
              >
                <div
                  class="card-body"
                >
                  <h3
                    class="title title-base"
                  >
                    Navigating Parenthood: Tips for New Moms
                  </h3>
                </div>
                <div
                  class="card-footer card-footer-light flex justify-between items-center"
                >
                  <div
                    class="writer-info grid"
                  >
                    <div
                      class="info-avatar"
                    >
                      <img
                        alt=""
                        class="object-fit-cover"
                        src="author3.jpg"
                      />
                    </div>
                    <div
                      class="info-intro"
                    >
                      <p
                        class="intro-name text-base"
                      >
                        Sophia Greene
                      </p>
                      <div
                        class="intro-verify flex items-center"
                      >
                        <img
                          alt="icon"
                          class="verify-icon"
                          src="verify_icon_filled.svg"
                        />
                        <span
                          class="verify-status text-sm text"
                        >
                          Verified writer
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    class="date-info text text-base"
                  >
                    20 August
                  </div>
                </div>
              </div>
            </div>
          </a>
          <a
            data-testid="slugLink"
            href="/blogs/post4"
          >
            <div
              class="card card-sm"
              data-testid="artice"
            >
              <img
                alt="Article"
                class="card-overlay-img object-fit-cover"
                data-testid="Image"
                src="blogImg4.jpg"
              />
              <div
                class="card-content"
              >
                <div
                  class="card-body"
                >
                  <h3
                    class="title title-base"
                  >
                    Embracing AI in the Workplace
                  </h3>
                </div>
                <div
                  class="card-footer card-footer-light flex justify-between items-center"
                >
                  <div
                    class="writer-info grid"
                  >
                    <div
                      class="info-avatar"
                    >
                      <img
                        alt=""
                        class="object-fit-cover"
                        src="author4.jpg"
                      />
                    </div>
                    <div
                      class="info-intro"
                    >
                      <p
                        class="intro-name text-base"
                      >
                        Alexander West
                      </p>
                      <div
                        class="intro-verify flex items-center"
                      >
                        <img
                          alt="icon"
                          class="verify-icon"
                          src="verify_icon_filled.svg"
                        />
                        <span
                          class="verify-status text-sm text"
                        >
                          Verified writer
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    class="date-info text text-base"
                  >
                    22 August
                  </div>
                </div>
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
